More magic

btel opened the pull request:
:mage_woman: Magic is all around so lets add more magic! :mage: 

For the list of supported magics see:

https://docs.google.com/spreadsheets/d/1xcBlx8yrNJIrgSQaAbiphr6dNlwd6Dw3a5bzi-Qa9yY/edit?usp=sharing

See also the notebook with examples:

https://github.com/btel/xeus-python/blob/more-magic/notebooks/xeus-magics.ipynb

# Screenshots

## History

![peek-history](https://user-images.githubusercontent.com/41565/83307363-d8421600-a204-11ea-84f4-1942c8b55d4d.gif)

## Recall

![peek-recall](https://user-images.githubusercontent.com/41565/83307389-e55f0500-a204-11ea-9705-db1299f61b73.gif)


## Timeit
![Peek-timeit](https://user-images.githubusercontent.com/41565/83300623-9f03a900-a1f8-11ea-89bf-accff362bcb8.gif)


## Quick help

![Peek 2020-05-28 23-54](https://user-images.githubusercontent.com/41565/83300619-9d39e580-a1f8-11ea-9018-2c283965e737.gif)

JohanMabille commented on src/xcomm.cpp:
```diff
@@ -187,6 +188,44 @@ namespace xpyt
         py::module kernel_module = py::module("kernel");
 
         py::class_<detail::xmock_object> _Mock(kernel_module, "_Mock");
+        py::class_<XInteractiveShell> XInteractiveShell(
```
Naming convention:

- we use snake case in C++ (even for classes), so this line should probably changed into `py::class_<xinteractive_shell> interactive_shell(// same as before)

- names starting with `_` or `__` are reserved for use by the standard library.

btel replied:
:+1:  but what do you do when you create a Python class wrapper for a C++ class? I guess you would use CamelCase for Python (like in the line above)?

JohanMabille commented on src/xinteractiveshell.cpp:
```diff
@@ -0,0 +1,119 @@
+    }
+
+
+    XInteractiveShell::XInteractiveShell() {
```
Formatting: opening curly brace should be on a new line.

JohanMabille commented on src/xinteractiveshell.hpp:
```diff
@@ -0,0 +1,73 @@
+
+    struct hooks_object
+    {
+        hooks_object() {}
```
This can ben removed, the compiler will generate it automatically.

JohanMabille commented on src/xinteractiveshell.hpp:
```diff
@@ -0,0 +1,73 @@
+
+    };
+
+    class XInteractiveShell {
```
Formatting: opening curly brace should be on a new line

JohanMabille commented on src/xinteractiveshell.hpp:
```diff
@@ -0,0 +1,73 @@
+        py::module _ipy_process;
+        py::module _magic_core;
+        py::module _magics_module;
+        py::module _extension_module;
```
Naming convention: we usually use `m_xxx` for attributes names.

The private section should come after the public one (most of the time you want to read the API, not the implementation details)

JohanMabille commented on src/xinteractiveshell.hpp:
```diff
@@ -0,0 +1,73 @@
+        py::object magics_manager;
+        py::object extension_manager;
+
+        hooks_object hooks;
```
Do we need write access for all of these properties? If not, they should be moved to the private section and getters should be provided. You can bind then the getters with `def_property_readonly`.

JohanMabille commented on src/xinteractiveshell.hpp:
```diff
@@ -0,0 +1,73 @@
+        void register_post_execute(py::args, py::kwargs) {};
+        void enable_gui(py::args, py::kwargs) {};
+        void observe(py::args, py::kwargs) {};
+        void showtraceback(py::args, py::kwargs) {};
```
I guess these are required by some python code?

btel replied:
yes ;-)

JohanMabille commented on src/xinteractiveshell.hpp:
```diff
@@ -0,0 +1,73 @@
+        void observe(py::args, py::kwargs) {};
+        void showtraceback(py::args, py::kwargs) {};
+
+        void init_magics();
```
This should be moved to the private section

btel commented:
@martinRenou when I change `p_history_manger` to `xeus:xhistory_manager &` and use `p_history_manager=history_manager` in the constructor, I get this error:

```
/home/bartosz/repos/libraries/xeus-python/src/xinteractiveshell.cpp:68:29: error: use of deleted function ‘xeus::xhistory_manager& xeus::xhistory_manager::operator=(const xeus::xhistory_manager&)’
   68 |         p_history_manager = history_manager;
      |                             ^~~~~~~~~~~~~~~
In file included from /home/bartosz/repos/libraries/xeus-python/src/xinteractiveshell.hpp:4,
                 from /home/bartosz/repos/libraries/xeus-python/src/xinteractiveshell.cpp:1:
/home/bartosz/.pyenv/versions/miniconda3-latest/envs/xeus/include/xeus/xhistory_manager.hpp:32:27: note: declared here
   32 |         xhistory_manager& operator=(const xhistory_manager&) = delete;
```

it seems that someone took care to delete the = operator for references.

JohanMabille commented:
@btel could you rebase instead of merging to avoid polluting history? Also squashing commits that are not meaningful with the meaningful ones?

btel commented:
Hi @JohanMabille, sorry I naively thought that it was what the github would do when I click "resolve conflicts" ;-). I squashed the commits into one and rebased them with master.

JohanMabille commented:
No worries ;) Thanks for rebasing, I'm gonna review it asap!

JohanMabille commented:
The issue on Windows is unrelated to thie PR. See https://bugs.python.org/issue36020 and https://github.com/nlohmann/json/issues/1408.

JohanMabille commented:
Thanks a lot!

btel commented:
Thanks, @JohanMabille . It was lighting fast! :heart_eyes:

JohanMabille commented:
;) I forgot to ask, why do we need the `py::dynamic_attr()` argument in the interactive shell binding?

btel commented:
hmm, maybe we don't need it anymore. I used that for testing when I was adding attributes on runtime, but I think now all of them are statically defined.
